builtin/providers/aws/resource_aws_instance.go

```diff
@@ -30,9 +30,17 @@ func resourceAwsInstance() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"ami": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"ami_name"},
+			},
+
+			"ami_name": &schema.Schema{
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"ami"},
```

Comment:
I agree reusing `ami` is better here (and leaving it required) - with a simple detection of ID vs string by whether or not it's prefixed with `ami`.